src/autoPagination.ts

```diff
@@ -4,82 +4,68 @@ import {callbackifyPromiseWithTimeout, getDataFromArgs} from './utils.js';
 };
-type IterationResult = {
-  done: boolean;
-  value?: any;
-};
+type IterationResult<T> =
+  | {
+      done: false;
+      value: T;
+    }
+  | {done: true; value?: undefined};
 type IterationDoneCallback = () => void;
-type IterationItemCallback = (
-  item: any,
+type IterationItemCallback<T> = (
+  item: T,
   next: any
 ) => void | boolean | Promise<void | boolean>;
-type ListResult = {
-  data: Array<any>;
-  // eslint-disable-next-line camelcase
-  has_more: boolean;
-};
-type AutoPagingEach = (
-  onItem: IterationItemCallback,
+type AutoPagingEach<T> = (
+  onItem: IterationItemCallback<T>,
   onDone?: IterationDoneCallback
 ) => Promise<void>;
 
 type AutoPagingToArrayOptions = {
   limit?: number;
 };
-type AutoPagingToArray = (
+type AutoPagingToArray<T> = (
   opts: AutoPagingToArrayOptions,
   onDone: IterationDoneCallback
-) => Promise<Array<any>>;
+) => Promise<Array<T>>;
 
-type AutoPaginationMethods = {
-  autoPagingEach: AutoPagingEach;
-  autoPagingToArray: AutoPagingToArray;
-  next: () => Promise<void>;
+type AutoPaginationMethods<T> = {
+  autoPagingEach: AutoPagingEach<T>;
+  autoPagingToArray: AutoPagingToArray<T>;
+  next: () => Promise<IterationResult<T>>;
   return: () => void;
 };
-
-export function makeAutoPaginationMethods(
-  self: StripeResourceObject,
-  requestArgs: RequestArgs,
-  spec: MethodSpec,
-  firstPagePromise: Promise<any>
-): AutoPaginationMethods {
-  const promiseCache: PromiseCache = {currentPromise: null};
-  const reverseIteration = isReverseIteration(requestArgs);
-  let pagePromise = firstPagePromise;
-  let i = 0;
-
-  // Search and List methods iterate differently.
-  // Search relies on a `next_page` token and can only iterate in one direction.
-  // List relies on either an `ending_before` or `starting_after` field with
-  // an item ID to paginate and is bi-directional.
-  //
-  // Please note: spec.methodType === 'search' is beta functionality and is
-  // subject to change/removal at any time.
-  let getNextPagePromise: (pageResult: any) => Promise<any>;
-  if (spec.methodType === 'search') {
-    getNextPagePromise = (pageResult): Promise<any> => {
-      if (!pageResult.next_page) {
-        throw Error(
-          'Unexpected: Stripe API response does not have a well-formed `next_page` field, but `has_more` was true.'
-        );
-      }
-      return self._makeRequest(requestArgs, spec, {
-        page: pageResult.next_page,
-      });
-    };
-  } else {
-    getNextPagePromise = (pageResult): Promise<any> => {
-      const lastId = getLastId(pageResult, reverseIteration);
-      return self._makeRequest(requestArgs, spec, {
-        [reverseIteration ? 'ending_before' : 'starting_after']: lastId,
-      });
-    };
+interface IStripeIterator<T> {
+  next: () => Promise<IterationResult<T>>;
+}
+type PageResult<T> = {
+  data: Array<T>;
+  has_more: boolean;
+  next_page: string | null;
+};
+class StripeIterator<T> implements IStripeIterator<T> {
+  private i: number;
```

Comment:
now that it's a field
```suggestion
  private index: number;
```
?